server/sandbox_run_linux.go

```diff
@@ -37,6 +39,234 @@ import (
 	"k8s.io/kubernetes/pkg/kubelet/types"
 )
 
+// addToMappingsIfMissing ensures the specified id is mapped from the host.
+func addToMappingsIfMissing(ids []idtools.IDMap, id int64) []idtools.IDMap {
+	firstAvailable := int(0)
+	for _, r := range ids {
+		if int(id) >= r.HostID && int(id) < r.HostID+r.Size {
+			// Already present, nothing to do
+			return ids
+		}
+		if r.ContainerID+r.Size > firstAvailable {
+			firstAvailable = r.ContainerID + r.Size
+		}
+	}
+	newMapping := idtools.IDMap{
+		ContainerID: firstAvailable,
+		HostID:      int(id),
+		Size:        1,
+	}
+	return append(ids, newMapping)
+}
+
+func (s *Server) configureSandboxIDMappings(mode string, sc *pb.LinuxSandboxSecurityContext) (*storage.IDMappingOptions, error) {
+	if mode == "" {
+		// No mode specified but mappings set in the config file, let's use them.
+		if s.defaultIDMappings != nil {
+			uids := s.defaultIDMappings.UIDs()
+			gids := s.defaultIDMappings.GIDs()
+			return &storage.IDMappingOptions{UIDMap: uids, GIDMap: gids}, nil
+		}
+		return nil, nil
+	}
+
+	// Ignore the annotation if not explicitly set in the config file.
+	if !s.config.AllowUsernsAnnotation {
```

Comment:
Can we move this as the first check for the function or out into the calling code?